src/cpp/lpnamer/problem_modifier/ProblemModifier.cpp

```diff
@@ -25,26 +15,22 @@ std::set<int> extract_time_steps(
   return result;
 }
 
-std::vector<Candidate> candidates_with_not_null_profile(
+std::vector<Candidate> ProblemModifier::candidates_with_not_null_profile(
     const std::vector<ActiveLink> &active_links,
-    const std::set<int> &time_steps) {
-  std::vector<Candidate> candidates = candidates_from_all_links(active_links);
-
-  candidates.erase(std::remove_if(candidates.begin(), candidates.end(),
-                                  [time_steps](const Candidate &cand) {
-                                    bool hasOnlyNullProfile = true;
-                                    for (int time_step : time_steps) {
-                                      hasOnlyNullProfile &=
-                                          cand.direct_profile(time_step) == 0.0;
-                                      hasOnlyNullProfile &=
-                                          cand.indirect_profile(time_step) ==
-                                          0.0;
-                                    }
-                                    return hasOnlyNullProfile;
-                                  }),
-                   candidates.end());
-
-  return candidates;
+    const std::set<int> &time_steps) const {
+  std::vector<Candidate> candidates_to_keep;
+
+  for (const auto& link: active_links) {
+    auto candidates = link.getCandidates();
+    auto newEnd = std::remove_if(
+        candidates.begin(), candidates.end(),
+        [&](const Candidate &cand) {
+          return cand.hasNullProfile(chronicleToUse(link), time_steps);
+        });
+
+    std::move(candidates.begin(), newEnd, std::back_inserter(candidates_to_keep));
+  }
+  return candidates_to_keep;
```

Comment:
If my understanding is correct, you

1. Run a loop through active links
2. For each active link, you select candidates that satisfy a `predicate`
3. Concatenate all the candidates, then `return` them

It seems that you create a copy of the candidates at l.24. You could probably do something like
```cpp
std::vector<Candidate> candidates_to_keep;
for (const auto link& : active_links) // you may use std::for_each
   for (const auto candidate& : link.getCandidates()) // I assume that getCandidates returns a const-ref
       if (predicate(candidate)) {
           candidates_to_keep.push_back(candidate);
       }
   }
   return candidates_to_keep;
}
```
Possibly keep `predicate` in a lambda-function. IM your intention would look more clear than using `<algorithm>` dark magic.

Reply:
"Black magic" was already there. I kept the usage of std::algorithm but I used your algorithm instead of the one in place.
The one in place: copy all candidates, remove those with null profile.
Your algorithm: look at all candidates, copy (push back) if not null profile
=> std::copy_if